### PR TITLE
ggml : fix mul_mat src1 indexing when src1 is not contiguous

### LINKED ISSUE #385: Odd behaviour with GPT2 after bfc6d42
New behaviour after bfc6d42f8c2141383e4f21e4a030688c71560da0

      $ ./bin/gpt-2 -m ./models/gpt-2-117M/ggml-model.bin -p 'The meaning of life is:' --top_k 1
      main: seed = 1689160875
      gpt2_model_load: loading model from './models/gpt-2-117M/ggml-model.bin'
      gpt2_model_load: n_vocab = 50257
      gpt2_model_load: n_ctx   = 1024
      gpt2_model_load: n_embd  = 768
      gpt2_model_load: n_head  = 12
      gpt2_model_load: n_layer = 12
      gpt2_model_load: ftype   = 1
      gpt2_model_load: qntvr   = 0
      gpt2_model_load: ggml tensor size = 240 bytes
      gpt2_model_load: ggml ctx size = 384.77 MB
      gpt2_model_load: memory size =    72.00 MB, n_mem = 12288
      gpt2_model_load: model size  =   239.08 MB
      extract_tests_from_file : No test file found.
      test_gpt_tokenizer : 0 tests failed out of 0 tests.
      main: prompt: 'The meaning of life is:'
      main: number of tokens in prompt = 6, first 8 tokens: 464 3616 286 1204 318 25 
      
      The meaning of life is: life
      
      
      "I'm
      
      "I'm
      
      "I'm
      
      "I'm

Old behaviour at bfc6d42f8c2141383e4f21e4a030688c71560da0^

    $ ./bin/gpt-2 -m ../../ggml/build/models/gpt-2-117M/ggml-model.bin -p 'The meaning of life is:' --top_k 1
    main: seed = 1689160978
    gpt2_model_load: loading model from '../../ggml/build/models/gpt-2-117M/ggml-model.bin'
    gpt2_model_load: n_vocab = 50257
    gpt2_model_load: n_ctx   = 1024
    gpt2_model_load: n_embd  = 768
    gpt2_model_load: n_head  = 12
    gpt2_model_load: n_layer = 12
    gpt2_model_load: ftype   = 1
    gpt2_model_load: qntvr   = 0
    gpt2_model_load: ggml tensor size = 240 bytes
    gpt2_model_load: ggml ctx size = 384.77 MB
    gpt2_model_load: memory size =    72.00 MB, n_mem = 12288
    gpt2_model_load: model size  =   239.08 MB
    extract_tests_from_file : No test file found.
    test_gpt_tokenizer : 0 tests failed out of 0 tests.
    main: prompt: 'The meaning of life is:'
    main: number of tokens in prompt = 6, first 8 tokens: 464 3616 286 1204 318 25 
    
    The meaning of life is: to live in a world of abundance, and to live in a world of abundance.

I haven't been able to figure out exactly why this happens, but I did bisect it down to that commit.  One thing I noticed when poking around in the debugger is that the logits for the first predicted token are correct, but the logits for the second predicted token differ.

### PR DESCRIPTION
**Pull request by ggerganov**
fix #385 

Sample run before fix:

<details>

```bash
$ ▶ ./bin/gpt-2 -m models/gpt-2-117M/ggml-model-f32.bin -p 'The meaning of life is:' -s 12 --top_k 1
main: seed = 12
gpt2_model_load: loading model from 'models/gpt-2-117M/ggml-model-f32.bin'
gpt2_model_load: n_vocab = 50257
gpt2_model_load: n_ctx   = 1024
gpt2_model_load: n_embd  = 768
gpt2_model_load: n_head  = 12
gpt2_model_load: n_layer = 12
gpt2_model_load: ftype   = 0
gpt2_model_load: qntvr   = 0
gpt2_model_load: ggml tensor size = 240 bytes
gpt2_model_load: ggml ctx size = 694.01 MB
gpt2_model_load: memory size =    72.00 MB, n_mem = 12288
gpt2_model_load: model size  =   474.70 MB
extract_tests_from_file : No test file found.
test_gpt_tokenizer : 0 tests failed out of 0 tests.
main: prompt: 'The meaning of life is:'
main: number of tokens in prompt = 6, first 8 tokens: 464 3616 286 1204 318 25 

The meaning of life is: life


"I'm

"I'm

"I'm

"I'm

"I'm

"I'm

"I'm

"I'm

"I'm

"I'm

"I'm

"I'm

"I'm

"I'm

"I'm

"I'm

"I'm

"I'm

"I'm

"I'm

"I'm

"I'm

"I'm

"I'm

"I'm

"I'm

"I'm

"I'm

"I'm

"I'm

"I'm

"I'm

"I'm

"I'm

"I'm

"I'm

"I'm

"I'm

"I'm

"

main: mem per token =  2010732 bytes
main:     load time =   194.41 ms
main:   sample time =    51.89 ms
main:  predict time =  2533.62 ms / 12.36 ms per token
main:    total time =  2810.71 ms
```
</details>

Sample run after fix:

<details>

```bash
$ ▶ ./bin/gpt-2 -m models/gpt-2-117M/ggml-model-f32.bin -p 'The meaning of life is:' -s 12 --top_k 1
main: seed = 12
gpt2_model_load: loading model from 'models/gpt-2-117M/ggml-model-f32.bin'
gpt2_model_load: n_vocab = 50257
gpt2_model_load: n_ctx   = 1024
gpt2_model_load: n_embd  = 768
gpt2_model_load: n_head  = 12
gpt2_model_load: n_layer = 12
gpt2_model_load: ftype   = 0
gpt2_model_load: qntvr   = 0
gpt2_model_load: ggml tensor size = 240 bytes
gpt2_model_load: ggml ctx size = 694.01 MB
gpt2_model_load: memory size =    72.00 MB, n_mem = 12288
gpt2_model_load: model size  =   474.70 MB
extract_tests_from_file : No test file found.
test_gpt_tokenizer : 0 tests failed out of 0 tests.
main: prompt: 'The meaning of life is:'
main: number of tokens in prompt = 6, first 8 tokens: 464 3616 286 1204 318 25 

The meaning of life is: to live in a world of abundance, and to live in a world of abundance.

The meaning of life is: to live in a world of abundance, and to live in a world of abundance.

The meaning of life is: to live in a world of abundance, and to live in a world of abundance.

The meaning of life is: to live in a world of abundance, and to live in a world of abundance.

The meaning of life is: to live in a world of abundance, and to live in a world of abundance.

The meaning of life is: to live in a world of abundance, and to live in a world of abundance.

The meaning of life is: to live in a world of abundance, and to live in a world of abundance.

The meaning of life is: to live in a world of abundance, and to live in a world of abundance.

The meaning of life is:

main: mem per token =  2010732 bytes
main:     load time =   190.43 ms
main:   sample time =    50.85 ms
main:  predict time =  2490.53 ms / 12.15 ms per token
main:    total time =  2761.98 ms
```
</details>

### CODE REVIEW

**Comment by ggerganov** [comment]
cc @sw for extra eyes on this change

File: src/ggml.c [hunk 1]
```diff
@@ -10828,7 +10830,15 @@ static void ggml_compute_forward_mul_mat(
...
 
         const char * src0_row = (const char *) src0->data + (  0 + i02*nb02 + i03*nb03     );
-        const char * src1_col = (const char *)      wdata + (i11 + i12*ne11 + i13*ne12*ne11)*row_size;
+
+        // desc: when src1 is not a contiguous memory block we have to calculate the offset using the strides
+        //       if it is, then we have either copied the data to params->wdata and made it contiguous or we are using
+        //       the original src1 data pointer, so we should index using the indices directly
+        // TODO: this is a bit of a hack, we should probably have a better way to handle this
+        const char * src1_col = (const char *) wdata +
+            (src1_cont
```
Comment (ggerganov) [inline src/ggml.c:10839]: Oops, this should have been:

```c
(src1_cont || src1->type != vec_dot_type
```
